appmap/_implementation/recorder.py

```diff
@@ -0,0 +1,184 @@
+import logging
+import threading
+import traceback
+from abc import ABC, abstractmethod
+
+from appmap._implementation.utils import appmap_tls
+
+logger = logging.getLogger(__name__)
+
+_default_recorder = None
+
+
+class Recorder(ABC):
+    """
+    A base class for Recorders.
+
+    Note that the abstract methods have implementations for use by subclasses.
+    """
+
+    @property
+    @abstractmethod
+    def events(self):
+        return self._events
+
+    @property
+    @abstractmethod
+    def next_event_id(self):
+        self._next_event_id += 1
+        return self._next_event_id
```

Comment:
There's no need for a method decorated with `@abstractmethod` to throw an exception . Inheriting from `ABC` ensures that only concrete subclasses can be instantiated. 

I think it's important to force the choice between a `ThreadRecorder` and a `SharedRecorder`. However, the since the difference between the two is only the locking required to ensure that `SharedRecorder` is thread-safe, having them share implementations seems like the right way to go.

Both of these points are discussed here: https://docs.python.org/3/library/abc.html?highlight=abc#abc.abstractmethod .